### PR TITLE
Improve matching for additional cases

### PR DESCRIPTION
**Pull request by bk2204**
This PR introduces three new commits to fix some additional edge cases in our pattern matching. Each commit describes its changes and adds tests for them.

### CODE REVIEW

File: wildmatch.go [hunk 1]
```diff
@@ -183,7 +192,7 @@ func (w *Wildmatch) parseTokens(dirs []string) []token {
 			return tokens
 		}
 	}
-	components := w.parseTokensSimple(dirs[0:])
+	components := w.parseTokensSimple(dirs[0:endItem])
```
Comment (chrisd8088) [inline wildmatch.go:195]: I noticed first we could probably have dropped the `0:` in the previous round of edits, and then I began to wonder if `dirs` could just be mutated instead of managing the `endItem` index.  Here's what I came up with; see if this makes any sense to you:
```go
	var finalComponents []token

	if !w.gitattributes {
		trailingIsEmpty := len(dirs) > 1 && dirs[len(dirs)-1] == ""
		numNonEmptyDirs := len(dirs)
		if trailingIsEmpty {
			numNonEmptyDirs -= 1
		}
		if w.contents {
			finalComponents = []token{&trailingComponents{}}
			if trailingIsEmpty {
				// Strip off the trailing empty string.
				dirs = dirs[:numNonEmptyDirs]
			}
		}
		// If we have one component, ignoring trailing empty
		// components and we know that a directory is permissible....
		if numNonEmptyDirs == 1 && (trailingIsEmpty || w.contents) {
			// We don't have a slash in the middle, so this can go
			// anywhere in the hierarchy.  If there had been a slash
			// here, it would have been anchored at the root.
			rest := w.parseTokensSimple(dirs)
			tokens := append([]token{&unanchoredDirectory{
				Until: rest[0],
			}})
			// If we're not matching all contents, then do include
			// the empty component so we don't match
			// non-directories.
			if finalComponents == nil && len(rest) > 1 {
				finalComponents = rest[1:]
			}
			return append(tokens, finalComponents...)
		}
	}
	components := w.parseTokensSimple(dirs)
	return append(components, finalComponents...)
```

**Review by chrisd8088 — APPROVED** [review]
Thanks for pursuing this further; I find it surprisingly twisty to try to reason about all the edge cases.  I had a few thoughts, but nothing which could not be ignored.

File: wildmatch.go [hunk 2]
```diff
@@ -159,13 +159,22 @@ func (w *Wildmatch) parseTokens(dirs []string) []token {
...
+			offset = 1
+		}
 		if w.contents {
 			finalComponents = []token{&trailingComponents{}}
+			// Strip off the trailing empty string.
+			endItem = len(dirs) - offset
 		}
-		if len(dirs) == 2 {
+		// If we have one component, ignoring trailing empty
+		// components and we know that a directory is permissible….
```
[the remark this answers is not in the document]
Comment (bk2204) [reply wildmatch.go:176]: That was intentional.  I believe the use of UTF-8 for Go source files is obligatory and of course it's 2021, so using the Unicode character seemed fine.  I do appreciate you pointing it out, though.

File: wildmatch.go [hunk 3]
```diff
@@ -349,13 +358,14 @@ type token interface {
 // doubleStar is an implementation of the Token interface which greedily matches
 // one-or-more path components until a successor token.
 type doubleStar struct {
-	Until token
+	Until     token
+	EmptyPath bool
```
[the remark this answers is not in the document]
Comment (bk2204) [reply wildmatch.go:362]: My feeling on this was that implicitly the other member was “match until” and so “match empty path” seemed to correspond appropriately.  Since this will only be set at the end (hence the name of the wrapping class, `trailingComponents`), whether we can continue is equivalent to whether this matches.

So I think I'd like to leave this as it is.

File: wildmatch.go (hunk 1 above)
Comment (bk2204) [reply wildmatch.go:195]: I've gone ahead and adopted this with a co-author credit.
